### ISSUE TITLE
Loss of qualifiers in function parameters in Clang template instantiations

### ISSUE THREAD

**Issue by wilsonk** [Syniurge/Calypso#20]
Hello Elie,

It looks like the list.d example stopped working a couple weeks back. 

/home/wilsonk/Downloads/Calypso/tests/calypso/libstdc++/list/list.d(76): Error: template cpp.std.list.list!(int, *allocator!int).list.remove_if cannot deduce function from argument types !()(bool function(const(int))), candidates are:
/usr/lib/gcc/x86_64-linux-gnu/4.8/../../../../include/c++/4.8/bits/stl_list.h(1363):        cpp.std.list.list!(int, *allocator!int).list.remove_if(_Predicate)(_Predicate)

Just thought I would let you know in case you didn't see it. The qtdemo (along with my expanded version that I haven't pushed yet) and all other libstdc++ examples work the same after this weeks changes, though, so no regressions there.

**Comment by Syniurge** [comment]
Hi Kelly,

I started working on the cppaggregate branch, only showcase.d works for the time being but it should catch up quickly since this simplifies things a great deal (the cost being that we'll get different semantics for D classes and for C++ classes, but no more ambiguities between whether a class is value or reference based on C++ compilation options).

I'll have a look at list.d once I bring back vector, string, bitset, deque and map.

**Comment by wilsonk** [comment]
Ok, sounds good Elie.

**Comment by Syniurge** [comment]
The bug seems to come from DMD: the deduced type for _Predicate throws away the const qualifier in « bool function(const(int) value) », so the previously matching remove_if instance gets discarded too by TypeFunction::callMatch().

Removing const in single_digit() makes list.d compile and run. I'll track where does const disappear tomorrow.

**Comment by Syniurge** [comment]
Okay, it's not DMD but this line in Sema while sending the (correct) template argument to instantiate remove_if<...>:

``` cpp
    ArgType = Context.getCanonicalType(ArgType);
```

Prior to this call:

```
 (gdb) call ArgType.dump()
 PointerType 0xc0d86f0 'PointerType FunctionProtoType BuiltinType _Bool (*)(const BuiltinType int)'
 `-FunctionProtoType 0xc0d8680 'FunctionProtoType BuiltinType _Bool (const BuiltinType int)' cdecl
   |-BuiltinType 0xa33e520 'BuiltinType _Bool'
   `-QualType 0xa33e5a1 'const BuiltinType int' const
     `-BuiltinType 0xa33e5a0 'BuiltinType int'
```

Then:

```
 (gdb) call ArgType.dump()
 PointerType 0xc0d86c0 'PointerType FunctionProtoType BuiltinType _Bool (*)(BuiltinType int)'
 `-FunctionProtoType 0xc0d8640 'FunctionProtoType BuiltinType _Bool (BuiltinType int)' cdecl
   |-BuiltinType 0xa33e520 'BuiltinType _Bool'
   `-BuiltinType 0xa33e5a0 'BuiltinType int'
```

Clang bug or expected behavior? If it's a bug it's not causing too much harm in Clang which is probably why it's still here, and if it's expected behavior I can't understand why and accustoming the Calypso side to this strange behavior of getCanonicalType won't be easy. In any case I think we should stick with the workaround for now.

**Comment by Syniurge** [comment]
It's still in latest Clang 3.7 and it looks like stripping the parameters from qualifiers in canonical function type is sought after..:

http://clang.llvm.org/doxygen/classclang_1_1ASTContext.html#aebe1c0862e34693bddd4683b51558e35 (ASTContext::getCanonicalParamType)

**Comment by wilsonk** [comment]
Looks like just using the workaround is the way to go. I tried compiling the cppaggregate branch to test this and I ran into an error in mtype.c...I suppose you must have this fixed on your end??

diff --git a/dmd2/mtype.c b/dmd2/mtype.c
index 19fcc2b..d87862b 100644
--- a/dmd2/mtype.c
+++ b/dmd2/mtype.c
@@ -9797,7 +9797,7 @@ AggregateDeclaration *getAggregateSym(Type *t)

 TypeClass *isClassValue(Type *t)
 {
-    if (t->ty != Tclass) return false;
-    if (t->ty != Tclass) return NULL;
   TypeClass\* tc = (TypeClass*)t;
   return tc->byRef() ? NULL : tc;
  }

Anyways, it looks like things are back to about the same, for libstdc++ tests, as they were a couple weeks ago. I'll see if I can get around to testing some other libs. I can see a few errors in my qt5demo extended version, so some rewriting will be needed.

Is there anything else left to do on this branch or is it going to merge with master right away?
